launcher/src/main/java/org/apache/spark/launcher/SparkClassCommandBuilder.java

```diff
@@ -0,0 +1,155 @@
+    } else if (className.equals("org.apache.spark.executor.CoarseGrainedExecutorBackend")) {
+      javaOptsKeys.add("SPARK_JAVA_OPTS");
+      javaOptsKeys.add("SPARK_EXECUTOR_OPTS");
+      memKey = "SPARK_EXECUTOR_MEMORY";
+    } else if (className.equals("org.apache.spark.executor.MesosExecutorBackend")) {
+      javaOptsKeys.add("SPARK_EXECUTOR_OPTS");
+      memKey = "SPARK_EXECUTOR_MEMORY";
+    } else if (className.startsWith("org.apache.spark.tools.")) {
+      String sparkHome = getSparkHome();
+      File toolsDir = new File(join(File.separator, sparkHome, "tools", "target",
+        "scala-" + getScalaVersion()));
+      checkState(toolsDir.isDirectory(), "Cannot find tools build directory.");
+
+      Pattern re = Pattern.compile("spark-tools_.*\\.jar");
+      for (File f : toolsDir.listFiles()) {
+        if (re.matcher(f.getName()).matches()) {
+          extraClassPath = f.getAbsolutePath();
+          break;
+        }
+      }
+
+      checkState(extraClassPath != null,
+        "Failed to find Spark Tools Jar in %s.\n" +
+        "You need to run \"build/sbt tools/package\" before running %s.",
+        toolsDir.getAbsolutePath(), className);
+
+      javaOptsKeys.add("SPARK_JAVA_OPTS");
+    } else {
+      // Any classes not explicitly listed above are submitted using SparkSubmit.
+      return createSparkSubmitCommand(env);
+    }
+
+    List<String> cmd = buildJavaCommand(extraClassPath);
+    for (String key : javaOptsKeys) {
+      addOptionString(cmd, System.getenv(key));
+    }
+
+    String mem = firstNonEmpty(memKey != null ? System.getenv(memKey) : null, DEFAULT_MEM);
+    cmd.add("-Xms" + mem);
+    cmd.add("-Xmx" + mem);
+    addPermGenSizeOpt(cmd);
+    cmd.add(className);
+    cmd.addAll(classArgs);
+    return cmd;
+  }
+
+  private List<String> createSparkSubmitCommand(Map<String, String> env) throws IOException {
+    final List<String> sparkSubmitArgs = new ArrayList<String>();
+    final List<String> appArgs = new ArrayList<String>();
+
+    // Parse the command line and special-case the HELP command line argument, allowing it to be
+    // propagated to the app being launched.
+    SparkSubmitOptionParser parser = new SparkSubmitOptionParser() {
+
+      @Override
+      protected boolean handle(String opt, String value) {
+        if (opt.equals(HELP)) {
+          appArgs.add(opt);
+        } else {
+          sparkSubmitArgs.add(opt);
+          sparkSubmitArgs.add(value);
+        }
+        return true;
+      }
+
+      @Override
+      protected boolean handleUnknown(String opt) {
+        appArgs.add(opt);
+        return true;
+      }
+
+      @Override
+      protected void handleExtraArgs(List<String> extra) {
+        appArgs.addAll(extra);
+      }
+
+    };
+
+    parser.parse(classArgs);
+    sparkSubmitArgs.add(parser.CLASS);
+    sparkSubmitArgs.add(className);
+
+    SparkSubmitCommandBuilder builder = new SparkSubmitCommandBuilder(true, sparkSubmitArgs);
```

Comment:
What I meant was that later in L150 we add the app arguments directly anyway instead of mixing it with the other Spark submit arguments, so we didn't need the `hasMixedArguments` there. IMO it didn't make sense to generalize spark class to mix arguments just for `HiveThriftServer2` since the existing spark class didn't do it either. Also it was weird for `SparkClassCommandBuilder` to call `SparkSubmitCommandBuilder` again since in the BASH script its the other way around.